### ISSUE TITLE
OSGi components not getting started when the dd-java-agent-0.67.0.jar added

### ISSUE THREAD

**Issue by deelaka96** [DataDog/dd-trace-java#2090]
Hi All,

We are using OSGI based products in our environment and we wanted to add data dog agent.
We have added the latest **dd-java-agent-0.67.0.jar** configured in our product and we have observed that a few of our OSGi components not getting activated. Ex:- Components in **Unsatisfied** state

`208	Unsatisfied		org.wso2.carbon.registry.info			org.wso2.carbon.registry.info(bid=566)`
`215	Unsatisfied		org.wso2.carbon.registry.task			org.wso2.carbon.registry.task(bid=583`

Also, we have observed a few imports added to our OSGi bundles.
`org.glassfish.grizzly.filterchain; version="1.7.0"<tyrus-standalone-client_1.7.0.wso2v1 [721]>
 com.google.common.util.concurrent; version="20.0.0"<com.google.guava_20.0.0 [36]>
 org.springframework.beans.factory.support; version="3.2.9.wso2v1"<spring.framework_3.2.9.wso2v1 [702]>
 org.springframework.context.support; version="3.2.9.wso2v1"<spring.framework_3.2.9.wso2v1 [702]>
 org.springframework.web.context; version="3.2.9.wso2v1"<spring.framework_3.2.9.wso2v1 [702]>`

Also, we did not observe this issue when we configure the **older version** of the data dog java agent in our environment.

Can the data dog agent related to the above-mentioned behavior.

Highly appreciate your support on the issue soon.

Regards,
Deelaka

**Comment by deelaka96** [comment]
Recreated the issue in the repository as per the following comment
https://github.com/DataDog/datadog-agent/issues/6750#issuecomment-728856161

**Comment by deelaka96** [comment]
Hi All,

**Steps to reproduce the issue**

1. Download the WSO2 API manager 2.6.0 [1]
2. Extract the zip file and go to wso2am-2.6.0/bin/ directory 
3. Edit the wso2server.sh (Linux/Mac) or wso2server.bat(windows)
4. Add the following in the startup arguments.

 -javaagent:/<location of the jar>/dd-java-agent-0.67.0-new.jar \
    -Ddd.service.name=api-gateway \
    -Ddd.service.mapping="mysql:api-gateway-mysql-db,h2:api-gateway-h2-db,default:api-gateway-default-db,okhttp:api-gateway-security-handler" \
    -Ddd.trace.debug=true\

Please find the sample screenshot

<img width="1228" alt="Screen Shot 2020-11-18 at 12 23 04 PM" src="https://user-images.githubusercontent.com/19946562/99495552-d6c5ff00-2998-11eb-962c-82f8091448fe.png">

In Windows add the above parameter in the **set CMD_LINE_ARGS=** property in the wso2server.bat

5. Then start the wso2 API Manager by executing the following commands.
On Windows:  wso2server.bat --run
On Linux/Mac OS:  sh wso2server.sh

You will be able to observe the mentioned behavior.

Also, the Equinox OSGi framework is used to build wso2 API manger.2.6.0.

Please refer to the following documentation on running the WSO2 API Manager.

[1]https://wso2.com/api-management/previous-releases/
[2]https://docs.wso2.com/display/AM260/Using+WSO2+API+Manager

Regards,
Deelaka

**Comment by mcculls** [comment]
Hi Deelaka, thanks for the detailed instructions - I'll take a look and post any updates here

**Comment by deelaka96** [comment]
Hi mcculls,

Any updates on this?

Regards,
Deelaka

**Comment by mcculls** [comment]
Hi Deelaka - I’m currently working on a solution, I'll update this issue when I have more details

**Comment by mcculls** [comment]
While I complete the fix I did manage to find a temporary workaround - if you set the following system property:
```
-Ddd.integration.guava.enabled=false
```
or the following environment variable:
```
DD_INTEGRATION_GUAVA_ENABLED=false
```
then the application should boot.

This will turn off instrumentation of some Guava types, but other instrumentation should still be enabled.

**Comment by deelaka96** [comment]
Hi mcculls,

Thank you for the update we will try the mentioned workaround until the complete fix is provided.

Regards,
Deelaka

**Comment by deelaka96** [comment]
Hi mcculls,
We have tried the workaround in our local environment and seem the issue got resolved.
Could you please briefly explain on what is causing this issue ? Guava library in Data dog jar?

**Comment by mcculls** [comment]
The root cause was that classpath probing by the tracer led to some optional Guava imports getting resolved before the Guava bundle was installed.

Optional imports are only ever resolved once in OSGi, unlike dynamic imports. If there's a bundle exporting that package at resolve time then the optional import will be wired to it and will use that export going forwards. However if there isn't any bundle exporting that package at resolve time then the optional import will be marked as not available. Any requests from that point onwards for the package will continue to fail, even if a bundle exporting that package is deployed later on.

The tracer does a little bit of classpath probing at boot to optimise what instrumentation matchers to apply. To do this it uses `getResource` to avoid loading classes or causing other side-effects. Unfortunately in OSGi even `getResource` can be enough to trigger bundle resolution.

In other words early probing by the tracer caused early resolution of certain WSO2 bundles before Guava was deployed, which meant that later on in the deployment they couldn't load Guava types that various services rely on. Turning off the Guava instrumentation removed probing of the Guava package, avoiding the early resolution.

The solution I've been working on is updating our matchers to try and detect this edge-case and avoid triggering the early bundle resolution.

PS. since these bundles appear to need Guava I do wonder if the current OSGi metadata is correct in marking them as optional

**Comment by deelaka96** [comment]
Thanks for the response mcculls but how come the older jar is working fine ?
any idea

**Comment by mcculls** [comment]
The earlier tracer jar had a bug involving classloader matching that meant that the Guava classpath checks were not being applied. Since these checks never happened, neither did the early package resolution. When the classloader matching was fixed in 0.64.0 (by #1920) the checks ran as expected and uncovered this issue.

Turning off the Guava instrumentation means the classloader checks involving Guava packages don't happen, which restores the previous behaviour (while still keeping the #1920 fix in place.)

**Comment by lpriima** [comment]
@deelaka96 could you please verify if you see the issue on latest release https://github.com/DataDog/dd-trace-java/releases/tag/v0.73.0 ?
